src/libzfs_core/py_zfs_core_module.c

```diff
@@ -1,17 +1,506 @@
 #include "../truenas_pylibzfs.h"
 
+static
+PyObject *PyExc_ZFSCoreError;
+
+PyDoc_STRVAR(py_zfs_core_exception__doc__,
+"ZFSCoreException(exception)\n"
+"-----------------------\n\n"
+"Python wrapper around error information returned by libzfs core\n\n"
+"attributes:\n"
+"-----------\n"
+"code: int\n"
+"    libzfs errno (one of ZFSError) or regular errno\n"
+"name: str\n"
+"    human-readable name of the libzfs errno\n"
+"errors: tuple | None\n"
+"    tuple of tuples containing error information returned by libzfs_core\n\n"
+"NOTE: the libzfs error may wrap around conventional OS errno. In this case\n"
+"it will be mapped to equivalent libzfs errno, but if that's not possible the\n"
+"libzfs errno will be set to EZFS_UNKNOWN and strerror output written to\n"
+"the error description field.\n"
+);
+PyObject *setup_zfs_core_exception(void)
+{
+	PyObject *dict = NULL;
+
+	dict = Py_BuildValue("{s:i,s:s,s:s,s:O}",
+			     "code", EZFS_UNKNOWN,
+			     "msg", "",
+			     "name", "",
+			     "errors", Py_None);
+	if (dict == NULL)
+		return NULL;
+
+	PyExc_ZFSCoreError = PyErr_NewExceptionWithDoc(PYLIBZFS_MODULE_NAME
+						       "lzc.ZFSCoreException",
+						       py_zfs_core_exception__doc__,
+						       PyExc_RuntimeError,
+						       dict);
+
+	Py_DECREF(dict);
+	return PyExc_ZFSCoreError;
+}
+
+static
+void set_zfscore_exc(const char *msg, int code, PyObject *errors_tuple)
+{
+	PyObject *v = NULL;
+	PyObject *args = NULL;
+	PyObject *attrs = NULL;
+	const char *name = NULL;
+	int err;
+
+	name = zfs_error_name(code);
+
+	v = PyObject_CallFunction(PyExc_ZFSCoreError, "s", msg);
+	if (v == NULL)
+		return;
+
+	attrs = Py_BuildValue(
+		"(issO)",
+		code,
+		msg,
+		name,
+		errors_tuple
+        );
+
+	if (attrs == NULL) {
+		Py_XDECREF(v);
```

Comment:
Why `Py_XDECREF` when `v` was already determined to be non null by this point?

Reply:
`v` is an allocated `ZFSCoreException`. `attrs` is NULL.